### PR TITLE
Fix allocateAoAs()

### PR DESCRIPTION
**Pull request by kishan-v**
#### Bug Fixes:
- fixed runtime error when submitting > 4 AoA preferences
- fixed Borda Count method

#### Features:
- randomly choose AoA in the event of a tie
  - Hence set base ExtendedAgent AoA preference to 0 (no preferences, happy with random)
  
#### Expected output
- if all agents have no AoA preference e.g. {0} then the AoA will be randomly assigned for each team, for each iteration of the game

 ```console
 -- Iteration 0 --
Team 9cedb286-4fcf-46d2-8523-8fc289404e8f has been assigned AoA 1
Team 40164d03-3381-4ea5-ae3a-ebe4c7fa31c3 has been assigned AoA 2
-- Iteration 1 --
Team f05fdf4a-96b0-49a5-936d-4ceb7f604710 has been assigned AoA 2
Team 9e1fb73c-6acf-449e-aebb-17f6b51580f5 has been assigned AoA 6
 ```

### CODE REVIEW

File: server/EnvironmentServer.go [hunk 1]
```diff
@@ -161,50 +166,86 @@ func (cs *EnvironmentServer) RunStartOfIteration(iteration int) {
...
-		var voteSum = []int{0, 0, 0, 0}
+		// Initialize a slice to hold weighted vote counts for AoAs 1 through 6.
+		// Index 0 is unused for simplicity.
+		voteSum := make([]int, 7) // voteSum[1] to voteSum[6]
+
+		// Iterate over each agent in the team
 		for _, agent := range team.Agents {
+			// Skip dead agents
 			if cs.IsAgentDead(agent) {
 				continue
```
Comment (nlewxxs) [inline server/EnvironmentServer.go:184]: Do dead agents remain in team.Agents? I thought they get removed?

File: server/EnvironmentServer.go [hunk 2]
```diff
@@ -161,50 +166,86 @@ func (cs *EnvironmentServer) RunStartOfIteration(iteration int) {
...
-		preference := 0
-		for aoa, voteCount := range voteSum {
-			if voteCount > currentMax {
-				currentMax = voteCount
-				preference = aoa
+		// Collect all AoAs that have the maximum vote count
+		var tiedAoAs []int
+		for aoa := 1; aoa <= 6; aoa++ {
+			if voteSum[aoa] == currentMax {
+				tiedAoAs = append(tiedAoAs, aoa)
```
Comment (nlewxxs) [inline server/EnvironmentServer.go:217]: nice

File: server/EnvironmentServer.go [hunk 3]
```diff
@@ -161,50 +166,86 @@ func (cs *EnvironmentServer) RunStartOfIteration(iteration int) {
...
 			}
-			for aoa, vote := range cs.GetAgentMap()[agent].GetAoARanking() {
-				voteSum[aoa] += vote
+
+			// Retrieve the agent's AoA rankings (up to 6 preferences)
+			aoaRanking := cs.GetAgentMap()[agent].GetAoARanking()
+
+			// Assign weights based on position
+			// First preference gets 6 points, second 5, ..., sixth 1
+			for position, aoa := range aoaRanking {
```
Comment (nlewxxs) [inline server/EnvironmentServer.go:192]: This is basically a borda count no? Or can be made into one using very little changes - can we modify it to make it clear that that's what we did here?

**Review by nlewxxs — CHANGES_REQUESTED** [review]

File: server/EnvironmentServer.go (hunk 3 above)
Comment (kishan-v) [reply server/EnvironmentServer.go:192]: Yep it is a Borda count. I added a comment just above (in #L169). I can try to make it clearer though.

File: server/EnvironmentServer.go (hunk 1 above)
Comment (kishan-v) [reply server/EnvironmentServer.go:217]: Yep good spot, those lines are unnecessary. Removed.

File: server/EnvironmentServer.go [hunk 4]
```diff
@@ -197,35 +201,65 @@ func (cs *EnvironmentServer) RunStartOfIteration(iteration int) {
...
 	for _, team := range cs.Teams {
-		// ranking cache for each team.
-		var voteSum = []int{0, 0, 0, 0}
+		// Array to store Borda Count totals for each AoA (1-6). Index 0 is unused for simplicity.
+		aoaVoteWeights := make([]int, 7) // aoaVoteWeights[1] to aoaVoteWeights[6]
+
+		// Process votes from all team members
 		for _, agent := range team.Agents {
-			if cs.IsAgentDead(agent) {
-				continue
```
Comment (kishan-v) [inline server/EnvironmentServer.go:210]: redundant code: dead agents are removed from `team.Agents`

File: server/EnvironmentServer.go [hunk 5]
```diff
@@ -239,10 +274,13 @@ func (cs *EnvironmentServer) allocateAoAs() {
 		case 6:
 			team.TeamAoA = common.CreateFixedAoA(1)
 		default:
+			// Default AoA assignment if no valid preference is found
 			team.TeamAoA = common.CreateFixedAoA(1)
```
Comment (johanjino) [inline server/EnvironmentServer.go:278]: Shouldnt we randomise from existing AoAs if we end up here?

Comment (johanjino) [reply server/EnvironmentServer.go:278]: Thanks solved by @kishan-v

**Review by johanjino — APPROVED** [review]

**Review by dhruvds12 — APPROVED** [review]

**Review by nlewxxs — APPROVED** [review]
hell yeah!

**Pull request merged**